crates/kitsune-core/src/activitypub/mod.rs

```diff
@@ -63,6 +67,48 @@ async fn handle_mentions(
     Ok(())
 }
 
+async fn handle_custom_emojis(
+    db_conn: &mut AsyncPgConnection,
+    post_id: Uuid,
+    fetcher: &Fetcher,
+    tags: &[Tag],
+) -> Result<()> {
+    let emoji_iter = tags.iter().filter(|tag| tag.r#type == TagType::Emoji);
+
+    let emoji_count = emoji_iter.clone().count();
+    if emoji_count == 0 {
+        return Ok(());
+    }
+
+    let futures = emoji_iter.clone().filter_map(|emoji| {
+        emoji
+            .id
+            .as_ref()
+            .map(|remote_id| fetcher.fetch_emoji(remote_id))
+    });
+
+    let emojis = join_all(futures)
+        .await
+        .into_iter()
+        .collect::<Result<Vec<_>, _>>()?
+        .iter()
+        .zip(emoji_iter)
```

Comment:
This `.zip` here doesn't seem quite correct. If emoji were filtered in the `.filter_map` step above, there would be a mismatch between the left and right values.

Consider passing them down inside a tuple from the future.

Reply:
That should also make it possible to get rid of the intermediate collect